src/backend/distributed/commands/truncate.c

```diff
@@ -216,13 +216,13 @@ EnsureLocalTableCanBeTruncated(Oid relationId)
 								  "tables.")));
 	}
 
-	/* make sure there are no foreign key references from a local table */
-	SetForeignConstraintRelationshipGraphInvalid();
-	List *referencingRelationList = ReferencingRelationIdList(relationId);
-
-	Oid referencingRelation = InvalidOid;
-	foreach_oid(referencingRelation, referencingRelationList)
+	List *referencingForeignConstaintsFromLocalTables =
+		GetForeignKeysFromLocalTables(relationId);
+	Oid foreignKeyId = InvalidOid;
+	foreach_oid(foreignKeyId, referencingForeignConstaintsFromLocalTables)
 	{
+		Oid referencingRelation = GetReferencingTableId(foreignKeyId);
+
 		/* we do not truncate a table if there is a local table referencing it */
 		if (!IsCitusTable(referencingRelation))
```

Comment:
`GetForeignKeysFromLocalTables` already returns ther foreign keys from local tables, so I think we don't need to check table type again `IsCitusTable`:

Maybe we can do something like:

```c
List *referencingForeignConstaintsFromLocalTables =
		GetForeignKeysFromLocalTables(relationId);
if list_length(referencingForeignConstaintsFromLocalTables) > 0:
  // error for first table in the list   
```